pkg/ebpf/cached_conntrack.go

```diff
@@ -0,0 +1,164 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/DataDog/datadog-agent/pkg/network/netlink"
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	ct "github.com/florianl/go-conntrack"
+	"github.com/golang/groupcache/lru"
+	"golang.org/x/sys/unix"
+)
+
+type refCountedConntrack struct {
+	ctrk  netlink.Conntrack
+	count *uint32
+}
+
+type cachedConntrack struct {
+	sync.Mutex
+	procRoot         string
+	cache            *lru.Cache
+	conntrackCreator func(int) (netlink.Conntrack, error)
+	closed           bool
+}
+
+func newCachedConntrack(procRoot string, conntrackCreator func(int) (netlink.Conntrack, error), size int) *cachedConntrack {
+	cache := &cachedConntrack{
+		procRoot:         procRoot,
+		conntrackCreator: conntrackCreator,
+		cache:            lru.New(size),
+	}
+
+	cache.cache.OnEvicted = func(_ lru.Key, v interface{}) {
+		v.(refCountedConntrack).Decr()
+	}
+
+	return cache
+}
+
+func (cache *cachedConntrack) Close() error {
+	cache.Lock()
+	defer cache.Unlock()
+
+	cache.cache.Clear()
+	cache.closed = true
+	return nil
+}
+
+func (cache *cachedConntrack) Exists(c *ConnTuple) (bool, error) {
+	ctrk, err := cache.ensureConntrack(c.NetNS(), int(c.Pid()))
+	if err != nil {
+		return false, err
+	}
+	defer ctrk.Decr()
+
+	var protoNumber uint8 = unix.IPPROTO_UDP
+	if c.isTCP() {
+		protoNumber = unix.IPPROTO_TCP
+	}
+
+	srcAddr, dstAddr := util.NetIPFromAddress(c.SourceAddress()), util.NetIPFromAddress(c.DestAddress())
+	srcPort, dstPort := c.SourcePort(), c.DestPort()
+
+	conn := netlink.Con{
+		Con: ct.Con{
+			Origin: &ct.IPTuple{
+				Src: &srcAddr,
+				Dst: &dstAddr,
+				Proto: &ct.ProtoTuple{
+					Number:  &protoNumber,
+					SrcPort: &srcPort,
+					DstPort: &dstPort,
+				},
+			},
+		},
+	}
+
+	ok, err := ctrk.ctrk.Exists(&conn)
+	if err != nil {
+		log.Errorf("error while checking conntrack for connection %#v: %s", conn, err)
+		cache.removeConntrack(c.NetNS())
+		return false, err
+	}
+
+	if ok {
+		return ok, nil
+	}
+
+	conn.Reply = conn.Origin
+	conn.Origin = nil
+	ok, err = ctrk.ctrk.Exists(&conn)
+	if err != nil {
+		log.Errorf("error while checking conntrack for connection %#v: %s", conn, err)
+		cache.removeConntrack(c.NetNS())
```

Comment:
is there a particular error we're worried about here? It's not clear why `removeConntrack` is the right thing to do here.

Reply:
I have seen sequence number issues with netlink sockets before (getting out of sync), which is primarily what I am concerned about here.